### PR TITLE
Aliassad/lp 1783

### PR DESCRIPTION
**Pull request by aliassad**
### Description

[LP-1783](https://philanthropyu.atlassian.net/browse/LP-1783)

[philu-edx-theme PR](https://github.com/philanthropy-u/philu-edx-theme/pull/1404)

Added a Google Recaptcha Backend logic

### CODE REVIEW

File: lms/envs/production.py [hunk 1]
```diff
@@ -1154,3 +1156,8 @@
 ########################## Derive Any Derived Settings  #######################
 
 derive_settings(__name__)
+
+
+######################### Secret Key for Google reCaptcha #####################
+
+CAPTCHA_SECRET_KEY = ENV_TOKENS.get('CAPTCHA_SECRET_KEY', None)
```
Comment (mabdullahz) [inline lms/envs/production.py]: Let's move this under the already existing `Philu Specific Settings` section, putting it at the bottom of that section with a simple comment as is already there for multiple other applications.

File: lms/envs/production.py [hunk 2]
```diff
@@ -927,6 +927,8 @@
 API_ACCESS_MANAGER_EMAIL = ENV_TOKENS.get('API_ACCESS_MANAGER_EMAIL')
 API_ACCESS_FROM_EMAIL = ENV_TOKENS.get('API_ACCESS_FROM_EMAIL')
 
+
+
```
Comment (mabdullahz) [inline lms/envs/production.py]: Let's remove this change if it is not serving any specific purpose

File: openedx/core/djangoapps/user_api/views.py [hunk 3]
```diff
@@ -130,6 +136,25 @@ def post(self, request):
...
+        response = urllib2.urlopen(req)
+        result = json.load(response)
+        # End reCAPTCHA validation #
+
+        if not result['success']:
+            errors = {
+                'g_recaptcha': [{"user_message": "Invalid reCaptcha"}]
+            }
+            return JsonResponse(errors, status=400)
+
```
Comment (mabdullahz) [inline openedx/core/djangoapps/user_api/views.py]: This is a core file, in which we have tried to do very very minimal changes. Please have a look at [this](https://github.com/philanthropy-u/edx-platform/blob/72763424b62efabea3307cfb2a92f0b42faeab44/openedx/core/djangoapps/user_authn/views/register.py#L154). This is the form we have used for backend validations. 
I suggest you study how the validations are being done through this form, and make the captcha validation a part of it. Feel free to have a discussion on this any time.

File: openedx/core/djangoapps/user_api/views.py [hunk 4]
```diff
@@ -30,7 +35,8 @@
 from openedx.core.djangoapps.user_api.helpers import require_post_params, shim_student_view
 from openedx.core.djangoapps.user_api.models import UserPreference
 from openedx.core.djangoapps.user_api.preferences.api import get_country_time_zones, update_email_opt_in
-from openedx.core.djangoapps.user_api.serializers import CountryTimeZoneSerializer, UserPreferenceSerializer, UserSerializer
+from openedx.core.djangoapps.user_api.serializers import CountryTimeZoneSerializer, UserPreferenceSerializer, \
+    UserSerializer
```
Comment (mabdullahz) [inline openedx/core/djangoapps/user_api/views.py]: This is a core file, let's please revert this change.

File: openedx/core/djangoapps/user_api/views.py [hunk 5]
```diff
@@ -1,5 +1,10 @@
 """HTTP end-points for the User API. """
 
+import urllib
+import urllib2
+import json
+
+from django.conf import settings
```
Comment (mabdullahz) [inline openedx/core/djangoapps/user_api/views.py]: Please do not forget to remove these imports from here once you move the logic to forms.
Also:
1. The imports are not in order, when you import in the forms please take care of import order
2. Let's only import the functions from these modules that we are going to use in these files, instead of importing the whole module

File: openedx/core/djangoapps/user_api/views.py [hunk 6]
```diff
@@ -130,6 +136,25 @@ def post(self, request):
...
+            'secret': settings.CAPTCHA_SECRET_KEY,
+            'response': recaptcha_response
+        }
+        data = urllib.urlencode(values)
+        req = urllib2.Request(url, data)
+        response = urllib2.urlopen(req)
+        result = json.load(response)
+        # End reCAPTCHA validation #
+
+        if not result['success']:
```
Comment (mabdullahz) [inline openedx/core/djangoapps/user_api/views.py]: If we only need to see the success, can't we rely on the response status code?

File: openedx/core/djangoapps/user_api/views.py [hunk 7]
```diff
@@ -130,6 +136,25 @@ def post(self, request):
...
 
+        # Begin reCAPTCHA validation #
+        recaptcha_response = data.get('g-recaptcha-response')
+        url = settings.CAPTCHA_VERIFY_URL
+        values = {
+            'secret': settings.CAPTCHA_SECRET_KEY,
+            'response': recaptcha_response
+        }
+        data = urllib.urlencode(values)
+        req = urllib2.Request(url, data)
```
Comment (mabdullahz) [inline openedx/core/djangoapps/user_api/views.py]: Let's consider using Python `requests` module here. [Reference](https://stackoverflow.com/questions/43324741/handle-google-recaptcha-with-python-requests)

File: lms/envs/production.py (hunk 1 above)
Comment (mabdullahz) [reply lms/envs/production.py]: Also, let's please not forget to prompt the DevOps that we need to add these tokens in production settings when regflow is deployed

Comment (mabdullahz) [reply lms/envs/production.py]: Also this should be moved to `auth.json` and used using `AUTH_TOKENS`

File: lms/envs/devstack_docker.py [hunk 8]
```diff
@@ -79,3 +79,7 @@
 # Uncomment the lines below if you'd like to see SQL statements in your devstack LMS log.
 # LOGGING['handlers']['console']['level'] = 'DEBUG'
 # LOGGING['loggers']['django.db.backends'] = {'handlers': ['console'], 'level': 'DEBUG', 'propagate': False}
+
+######################### Secret Key for Google reCaptcha #####################
+
+CAPTCHA_SECRET_KEY = ENV_TOKENS.get('CAPTCHA_SECRET_KEY', None)
```
Comment (mabdullahz) [inline lms/envs/devstack_docker.py]: `ENV_TOKENS` does not seem to be defined here?

File: openedx/core/djangoapps/user_api/views.py [hunk 9]
```diff
@@ -130,6 +136,25 @@ def post(self, request):
         email = data.get('email')
         username = data.get('username')
 
+        # Begin reCAPTCHA validation #
+        recaptcha_response = data.get('g-recaptcha-response')
+        url = settings.CAPTCHA_VERIFY_URL
```
Comment (ahmed-faraz46) [inline openedx/core/djangoapps/user_api/views.py]: Because it's not being use at multiple places lets please use it directly instead a making a separate `url` variable.

File: openedx/core/djangoapps/user_api/views.py [hunk 10]
```diff
@@ -130,6 +136,25 @@ def post(self, request):
         email = data.get('email')
         username = data.get('username')
 
+        # Begin reCAPTCHA validation #
```
Comment (ahmed-faraz46) [inline openedx/core/djangoapps/user_api/views.py]: After moving this validation in the form, let's please remove these comments.

File: lms/envs/common.py [hunk 11]
```diff
@@ -3552,3 +3552,6 @@ def _make_locale_paths(settings):
 # CDN LINK
 
 CDN_LINK = "https://static.philanthropyu.org/"
+
+# Google reCaptcha verify URL
+CAPTCHA_VERIFY_URL = "https://www.google.com/recaptcha/api/siteverify"
```
Comment (ahmed-faraz46) [inline lms/envs/common.py]: Let's please move this in student_accounts/constants, because for now, I can't foresee its being used anywhere else. So instead of making core change, we should move it to our app.

File: openedx/core/djangoapps/user_api/views.py (hunk 4 above)
Comment (aliassad) [reply openedx/core/djangoapps/user_api/views.py]: Done reverted

File: openedx/core/djangoapps/user_api/views.py (hunk 10 above)
Comment (aliassad) [reply openedx/core/djangoapps/user_api/views.py]: Done removed

File: lms/envs/production.py [hunk 12]
```diff
@@ -1124,9 +1124,15 @@
 # SurveyGizmo settings
 SURVEY_GIZMO_TOKEN = AUTH_TOKENS.get('SURVEY_GIZMO_TOKEN', None)
 SURVEY_GIZMO_TOKEN_SECRET = AUTH_TOKENS.get('SURVEY_GIZMO_TOKEN_SECRET', None)
-	# Django channels settings
+
+# Django channels settings
 CELERY_BROKER_HOSTNAME = ENV_TOKENS.get('CELERY_BROKER_HOSTNAME', None)
 
+# Settings for Google reCaptcha
+CAPTCHA_VERIFY_URL = "https://www.google.com/recaptcha/api/siteverify"
```
Comment (ahmed-faraz46) [inline lms/envs/production.py]: Can we consider moving all these settings in `common.py`? Because its more meaningful, all settings which are common between all of our servers are placed in common.

File: openedx/features/student_account/forms.py [hunk 13]
```diff
@@ -76,6 +84,21 @@ def clean_organization_name(self):
                 return org_name
         return org_name
 
+    def clean_g_recaptcha(self):
```
Comment (ahmed-faraz46) [inline openedx/features/student_account/forms.py]: Are we sure this cleaner is being called? becuase field name is not consistent with cleaner name.

File: openedx/features/student_account/forms.py [hunk 14]
```diff
@@ -76,6 +84,21 @@ def clean_organization_name(self):
...
+        g_recaptcha = self.cleaned_data.get('recaptcha')
```
Comment (ahmed-faraz46) [inline openedx/features/student_account/forms.py]: Let's please choose the same naming convention across the functionality either `g_recaptcha` or `recaptcha`

File: openedx/features/student_account/forms.py [hunk 15]
```diff
@@ -76,6 +84,21 @@ def clean_organization_name(self):
...
+        response = requests.post(
+            settings.CAPTCHA_VERIFY_URL,
+            data={
+                'secret': settings.CAPTCHA_SECRET_KEY,
+                'response': g_recaptcha
+            }).content
+        result = json.loads(response)
```
Comment (ahmed-faraz46) [inline openedx/features/student_account/forms.py]: can we please consider using some specific variable names instead of `response` and `result`.

File: lms/envs/production.py (hunk 12 above)
Comment (aliassad) [reply lms/envs/production.py]: AUTH_TOKENS was not available in common.py thats why i have added it here.

File: openedx/features/student_account/forms.py (hunk 15 above)
Comment (aliassad) [reply openedx/features/student_account/forms.py]: renamed  result to response_json making use of one generic name only.
